### PR TITLE
feat(eslint-plugin): [unbound-method] Support Array methods with 'this' argument

### PR DESCRIPTION
**Pull request by jacekjagiello**
Closes: #743

Hey, this is my first PR. I don't have much experience with AST, so it feels for me like walking in the fog 😄 I assume there is a better solution for this bug. Any help appreciated. I'll leave some questions in PR

### CODE REVIEW

File: packages/eslint-plugin/src/rules/unbound-method.ts [hunk 1]
```diff
@@ -108,6 +120,21 @@ function isSafeUse(node: TSESTree.Node): boolean {
       return true;
 
     case AST_NODE_TYPES.CallExpression:
+      if (
+        parent.callee.type === AST_NODE_TYPES.MemberExpression &&
+        parent.callee.property.type === AST_NODE_TYPES.Identifier
+      ) {
```
Comment (jacekjagiello) [inline packages/eslint-plugin/src/rules/unbound-method.ts]: What I really would like to do, is to check if callee for ex. `nodes` is an `Array`. Is this possible?

File: packages/eslint-plugin/src/rules/unbound-method.ts [hunk 2]
```diff
@@ -108,6 +120,21 @@ function isSafeUse(node: TSESTree.Node): boolean {
...
+        const method = parent.callee.property.name;
+        const args = parent.arguments;
+
+        if (Object.values(MethodsWithThisArg).includes(method)) {
```
Comment (jacekjagiello) [inline packages/eslint-plugin/src/rules/unbound-method.ts]: if there is other way to check this, let me know

File: packages/eslint-plugin/src/rules/unbound-method.ts (hunk 1 above)
Comment (bradzacher) [reply packages/eslint-plugin/src/rules/unbound-method.ts]: you should be able to get the type of the node, and ensure its type is an array.
I can't remember the flags exactly, so you might want to spend some time with the debugger in vscode.

File: packages/eslint-plugin/src/rules/unbound-method.ts (hunk 2 above)
Comment (bradzacher) [reply packages/eslint-plugin/src/rules/unbound-method.ts]:
```suggestion
        if (methodsWithThisArg.has(method)) {
```

File: packages/eslint-plugin/src/rules/unbound-method.ts [hunk 3]
```diff
@@ -134,3 +161,7 @@ function isSafeUse(node: TSESTree.Node): boolean {
 
   return false;
 }
+
+function getThisArgArgument(method: string, args: TSESTree.Expression[]) {
+  return method === MethodsWithThisArg.From ? args[2] : args[1];
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/unbound-method.ts]:
```suggestion
  return method === 'from' ? args[2] : args[1];
```

File: packages/eslint-plugin/src/rules/unbound-method.ts (hunk 1 above)
Comment (jacekjagiello) [reply packages/eslint-plugin/src/rules/unbound-method.ts]: I've spent some time trying to play with the debugger to find any flags that might indicate `Array` type, but with no success for now. If you think we can't leave it like this, I'll spend more time and maybe find the better way.

Comment (bradzacher) [reply packages/eslint-plugin/src/rules/unbound-method.ts]: Digging back into this, sorry a lot going on. Finally taken time to tackle the PR queue.

Feel free to add this file into your PR, and use that API.
https://github.com/typescript-eslint/typescript-eslint/blob/naming-conventions/packages/eslint-plugin/typings/typescript.d.ts

The API is "internal", but it's stable - they just haven't gotten around to exposing their type relationship API yet (there's a whole discussion around why they haven't yet...)

You can then pass the typechecker into this function and call `checker.isArrayType(type)` (you'll also have to pass the type map in).

Also just as an FYI - this site is amazing for playing with TS's type API: https://ts-ast-viewer.com/
You can open the chrome dev tools, and interrogate the selected type via the global variables it maintains: `type`, `program` and `typeChecker`.

**Pull request closed without merging**